apps/enc.c

```diff
@@ -418,15 +426,26 @@ int enc_main(int argc, char **argv)
                     BIO_printf(bio_err, "bad magic number\n");
                     goto end;
                 }
-
                 sptr = salt;
             }
 
-            if (!EVP_BytesToKey(cipher, dgst, sptr,
-                                (unsigned char *)str,
-                                str_len, 1, key, iv)) {
-                BIO_printf(bio_err, "EVP_BytesToKey failed\n");
-                goto end;
+            if (pbkdf2 == 1) {
+                if (!PKCS5_PBKDF2_HMAC(str, str_len, sptr, sizeof(salt), iter,
+                                       dgst,
+                                       EVP_CIPHER_key_length(cipher), key)) {
+                    BIO_printf(bio_err, "PKCS5_PBKDF2_HMAC failed\n");
+                    goto end;
+                }
+                /* should we reject if no IV is given ? */
```

Comment:
Can we not generate the IV in the PKCS5_PBKDF2_HMAC call above? i.e. instead of generating data that is "key length" long, we generate data that is "key length + iv length" long.